crates/openvino/src/request.rs

```diff
@@ -1,59 +1,222 @@
-use crate::blob::Blob;
+use crate::tensor::Tensor;
 use crate::{cstr, drop_using_function, try_unsafe, util::Result};
 use openvino_sys::{
-    ie_infer_request_free, ie_infer_request_get_blob, ie_infer_request_infer,
-    ie_infer_request_infer_async, ie_infer_request_set_batch, ie_infer_request_set_blob,
-    ie_infer_request_t, ie_infer_request_wait,
+    ov_callback_t, ov_infer_request_cancel, ov_infer_request_free,
+    ov_infer_request_get_input_tensor, ov_infer_request_get_input_tensor_by_index,
+    ov_infer_request_get_output_tensor, ov_infer_request_get_output_tensor_by_index,
+    ov_infer_request_get_tensor, ov_infer_request_infer, ov_infer_request_set_callback,
+    ov_infer_request_set_input_tensor, ov_infer_request_set_input_tensor_by_index,
+    ov_infer_request_set_output_tensor, ov_infer_request_set_output_tensor_by_index,
+    ov_infer_request_set_tensor, ov_infer_request_start_async, ov_infer_request_t,
+    ov_infer_request_wait, ov_infer_request_wait_for,
 };
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use std::time::Duration;
 
-/// See
-/// [`InferRequest`](https://docs.openvinotoolkit.org/latest/classInferenceEngine_1_1InferRequest.html).
+/// See [`InferRequest`](https://docs.openvino.ai/2024/api/c_cpp_api/group__ov__infer__request__c__api.html).
 pub struct InferRequest {
-    pub(crate) instance: *mut ie_infer_request_t,
+    pub(crate) instance: *mut ov_infer_request_t,
 }
-drop_using_function!(InferRequest, ie_infer_request_free);
+drop_using_function!(InferRequest, ov_infer_request_free);
 
 unsafe impl Send for InferRequest {}
 unsafe impl Sync for InferRequest {}
 
 impl InferRequest {
-    /// Set the batch size of the inference requests.
-    pub fn set_batch_size(&mut self, size: usize) -> Result<()> {
-        try_unsafe!(ie_infer_request_set_batch(self.instance, size))
+    // Tensors
+
+    /// Retrieve a [Tensor] from the input or output of the model.
+    pub fn tensor(&self, name: &str) -> Result<Tensor> {
+        let name = cstr!(name);
+        let mut tensor = std::ptr::null_mut();
+        try_unsafe!(ov_infer_request_get_tensor(
+            self.instance,
+            name.as_ptr(),
+            std::ptr::addr_of_mut!(tensor)
+        ))?;
+        Ok(Tensor { instance: tensor })
+    }
```

Comment:
Functions like this are problematic because the `Tensor` type assumes it has ownership over the internal `instance` pointer. When the `Tensor` goes out of scope, it's dropped and the underlying tensor memory is freed even though it still exists in the `InferRequest`.

Reply:
Nevermind, I checked the OpenVINO source code and everything uses C++ shared_ptr under the hood.